src/nfa/nfa.cc

```diff
@@ -473,8 +474,11 @@ Post& Delta::get_mutable_post(State q) {
     return posts[q];
 }
 
-std::vector<State> Delta::defragment(NumberPredicate<State>& is_staying) {
+void Delta::defragment(const std::vector<char>& is_staying, const std::vector<State>& renaming) {
+    //TODO: this function seems to be unreadable, should be refactored, maybe into several functions with a clear functionality?
+
     //first, indexes of post are filtered (places of to be removed states are taken by states on their right)
+    // Why would anyone write such horrible code?
```

Comment:
But the howle thing is big blob which is kind of impenetrable, maybe we could just write it a little more idiot friendly.

Reply:
We can. It will be slower. This is efficient implementation in C++. [Wiki](https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom), [random blog post](https://www.freecodecamp.org/news/how-to-remove-elements-from-a-container-in-c/).